zaino-state/src/chain_index.rs

```diff
@@ -20,3 +20,65 @@ pub mod mempool;
 pub mod non_finalised_state;
 /// Common types used by the rest of this module
 pub mod types;
+
+use std::{collections::HashMap, sync::Arc};
+
+use futures::Stream;
+use non_finalised_state::{BlockchainSource, NonfinalizedBlockCacheSnapshot};
+pub use zebra_chain::parameters::Network as ZebraNetwork;
+use zebra_state::HashOrHeight;
+
+/// The combined index. Contains a view of the mempool, and the full
+/// chain state, both finalized and non-finalized, to allow queries over
+/// the entire chain at once. Backed by a source of blocks, either
+/// a zebra ReadStateService (direct read access to a running
+/// zebrad's database) or a jsonRPC connection to a validator.
+///
+/// TODO: Currently only contains the non-finalized state.
+pub struct ChainIndex {
+    // TODO: finalized state
+    // TODO: mempool
+    non_finalized_state: non_finalised_state::NonFinalizedState,
+}
+
+/// The interface to the chain index
+pub trait Index: Sized {
+    /// Creates a new chainindex from a connection to a validator
+    /// Currently this is a ReadStateService or JsonRpSeeConnector
+    async fn new<T: Into<BlockchainSource>>(
+        source: T,
+        network: ZebraNetwork,
+    ) -> Result<Self, non_finalised_state::InitError>;
+    /// Takes a snapshot of the non_finalized state. All NFS-interfacing query
+    /// methods take a snapshot. The query will check the index
+    /// it existed at the moment the snapshot was taken.
+    fn snapshot_nonfinalized_state(&self) -> Arc<NonfinalizedBlockCacheSnapshot>;
```

Comment:
```suggestion
    fn snapshot_nonfinalized_state(&self) -> Arc<Self::Snapshot>;
```
Is this truly infallible? If not:
```suggestion
    fn snapshot_nonfinalized_state(&self) -> Result<Arc<Self::Snapshot>, Self::Error>;
```

Reply:
I'm not sure whether the concrete `Arc` should actually be exposed here; if instead, it returned just `Self::Snapshot` then in the implementation for the concrete type, you might choose `Self::Snapshot = Arc<NonfinalizedBlockCacheSnapshot>`. Then, you wouldn't need all of the `AsRef + Clone`.